README.md
# hanoitoys Package
last change 08-Feb-2022
### A collection of Tower of Hanoi solutions.
aka Tower of Lucas, 
aka Tower of Brahma, 
mostly toy programs showing non recursive solutions to the Classical (3 pegs/needles) Hanoi.  
One solution generator to Super Hanoi aka Hanoi on steroids.
This is the Tower of Hanoi with more than 3 pegs/needles. This is the Steward Frame algorithm.
This has an optional turtle grapics display.


## The toy solutions
### HanoiClassical.py
This is the classical Hanoi solution using recursion. The CS 101 method.  
#### 1st the classical Tower of Hanoi
The classical Tower of Hanoi has 3 posts or needles and n discs. One of the posts is where all n discs are located at when the puzzle starts. This post 
is called source.
The goal is to move all the discs to the destination peg. Subject to 3 rules.  
Rule # 1 : A larger disc is never placed on top of a smaller disc.  
Rule # 2 : Only one Disc may be moved at a time.  
Rule # 3 is that in that the sequence of moves leading to the solution may not have any repeated states.  
There is another disc labeled Store.   


This leads us to the fact there is only one sequence of states that leads to a solution for classical Tower of Hanoi.  

In a second sense this is the classical most known solution, the recursive solution.  
The base case is you can move one disk from x to y when y does not have any smaller disc.
Each level of recursion does relabel the pegs. The former labeling is restored by returning.
```
def recursive_hanoi (source, store, destination, n):
    if n == 1:
        move the one disc
        return
    recursive_hanoi (source, destination, store, n - 1
    recursive_hanoi (source, store, destination, 1)
    recursive_hanoi (store, source, destination, n - 1)
```
#### HanoiOlive.py is a nonrecursive Olive algorithm.
Francois Edourard Lucas who wrote about this problem and gave it the Tower of Hanoi name (writing under the nom de plume of M Claus, Claus being an anagram of Lucas) 
. Raoul Olive, a nephew of Lucas, observed the smallest
disc moves every other move. The motion of this peg depends on the parity of n.  
When n is odd the smallest disc moves from source to destination to store, then back to source.  
When n is even the smallest disc moves from source to store to destination, then back to source.  
When the smallest disc is not being moved there is only 1 legal move involving the two pegs the smallest disc is not occupying.
#### HanoiIdlePeg.py is another non recursive refinement.
Here we add a marker which is like a smaller disc which marks the Idle (non participating peg). With a real puzzle with discs and wooden pegs you can often
let a sewing thimble represent the Idle peg. We do not count the Idle Peg token in the number of discs n.  
For Odd n we move clockwise source -> store -> destination -> source.  
For Even n we move counter clockwise source -> destination -> store -> source.  
we repeatedly move the Idle token then find the legal move that can be made with the remaining participating pegs.

#### HanoiNR.py
HanoiNR.py is also included. This is the Wikipedia non recursive Tower of Hanoi solution.
## Tower of Hanoi on steroids. Number of pegs > 3. The Steward, Frame Algorithm.  CS 501
While Francois Edourard Lucas aka M Claus did undoubtedly write on the more than three peg Tower of Hanoi, we will now visit that problem disguised as the Reve's
Puzzle. In 1907 Henry Ernest Dudeney wrote a book of mathematical puzzles called "*The Canterbury Puzzles and Other Curious Problems*". In this work the characters are from Chaucer's "*The Canterbury Tales*" are on a pilgrimage and resting at a tavern where each of them presents a puzzle to their fellow travelers. The first of these is the Reve's[^1] puzzle. In the Reve's puzzle the are 4 stools and 8 cheeses of various sizes. These are on a stool in order, the smaller the cheese the closer to the top of the stack. The puzzle is to move the cheeses to another stool in the same order. Only one cheese can be moved at a time. A cheese can only be moved to an empty stool or the top of a larger cheese. Only one cheese can be moved at a time.

Clearly  this is equivalent to a 4 peg/needle 8 disc Tower of Hanoi problem.

### Reve's puzzle being solved by Dan's Hanoi (Hanoi.py in this package)

![Reve's puzzle being solved by Dan's Hanoi](Reves.png)

In 1941 B. M. Stewart and J. S. Frame each independently came up with
a algorithm to solve the generalized (n > 3) Tower of Hanoi puzzle. This
partitioned the initial stack of disks into sequential groups. The
subproblems being to move each group to a peg (a recursive task) and
reassemble them on the destination peg/needle.  They believed this
solution was 'Optimal'. This is known as the Frame Stewart conjecture.
In 2018 this conjecture was proved by Roberto Demontis.  
#### What is Optimal?
  In classical (3 peg/needle) Hanoi the constraint to not
repeat a state means there is only one solution. That solution is
optimal, there are no solutions using fewer moves because there are no
other solutions period. The optimal solution has 2^n^-1 solutions.
This is not true when you go beyond 3 pegs. There are many solutions,
even when we keep the no repeated state requirement. The ones where no
other solutions that take less moves are optimal. Clearly you can
remap all pegs/needles other than the source and destination in (n-2)!
ways. And at each sub problem you can remap. There are therefore many
Optimal solutions.  

#### The solutions take far fewer moves than the classical Tower of Hanoi does.
Consider the 64 disc 4 needle
solutions: You could just not use one needle and solve this
classically. Or you could move the top disc to a needle and solve the
rest as a 63 disc classical Tower of Hanoi. This would take a tiny bit
more than half the moves of the 64 disc classical Tower of Hanoi
solution. Or move two discs and then 62 classical, which would be a
fourth the moves. The first few n values of this are: 
|n|Number of moves| 
|---|------------------------| 
|0|18,446,744,073,709,551,615|
|1|9,223,372,036,854,775,809|
|2|4,611,686,018,427,387,909|
|3|2,305,843,009,213,693,965|
|4|1,152,921,504,606,847,005|
|5|576,460,752,303,423,549|
At n=53 we do the best we can at 18,433 moves.  

#### We need an estimate for cost.  
We can guide our algorithm by a Cost[^2] Estimator. 
It too will be recursive. We just have to
know how many moves, not actually make the moves. However the
estimator has to not only tell us how few moves we can get away with
but what direction we have to go to get there.  
Here is an estimator:
``` 
@functools.lru_cache(maxsize=3000) 
def est (NumDiscs: int, NumStore: int) -> tuple: 
    # returned tuple (num of moves, first move this many to storage needle 
    # can use min on list of these tuples to get tuple to return.  
    # or alternatively track the minimum without a list 
    if NumStore == 1 or NumDiscs < 3:
       # base case is classic Hanoi.
       return ((1 << NumDiscs) - 1, 1) # 2 ** NumDiscs - 1; this is faster.
    possibilities = list() 
    # build a list of sub problems 
    for n in range(1, NumDiscs):
	moves = est(n, NumStore)[0] << 1		# n src->stor, n
       stor->dest moves += est(NumDiscs - n, NumStore - 1)[0] # NumDiscs-n
       src->dest possibilities.append((moves, n))	
       # ok in list see if it is best later 
       return min(possibilities) 
```
This recursive solution is of order n factorial O(n!) Far worse for n > 3 than say Fibbinacci
O(2^n^). If you try this for yourself best use *memoization*.  The
lru_cache decorator from functools is a godsend. Fail to do this means
the mean time to fail for the hardware may foil your attempt.  
The estimator 'est' returns a tuple, the first element is the number of
moves (the thing we want to minimize), The second is the number of
discs from the top of the source peg/needle to move to one of the available storage
peg/needles.

 As to which available peg/needle it does not matter.
Initially this was coded with the Python set object to hold the
available storage pegs, but to make it seem as if it was not just
guessing, I implemented a ordered set which is just an int with bits
set for pegs.  Note: At a any particular point in the recursion some
of the available pegs may have discs on them, only they will be bigger
discs then the ones you are trying to move.  

#### Labeling the pegs/needles.

When there are more than 3 needles, calling them source, store and destination no longer works.
The Hanoi.py in the package names them first after the English alphabet Uppercase. The 27th is lower case 'a'.
We continue on through the lower case letter to the 52nd which is lower case zed 'z'. Needles 53 through 76 are given
Greek letters, which are spelled out in English in the normal output and use unicode equivalents in the graphics.


The source at the top level is always labeled 'A'. The destination is last needle, if it were the 64th it would be 'nu'.

#### The Dialogue.

The Dialogue is not a Turtle dialogue or a tkinter dialogue. This is a default stdin/stdout dialog. Turtle is not even initialized until
we know we need a graphical output. There are two flavors: text mode output and graphics. The graphics dialog needs the same information and
a little bit more than the text dialog. So the last question of the text mode dialogue is "Do you want a graphical representation?" to which
a reply of the single character 'n' is acceptable. The other questions are "Enter Number of Needles :" and
"Enter Number of Discs :".


The example shown sets up for a textual only solution to the Reve's puzzle.
```
Enter Number of Needles : 4
Enter Number of Discs : 8
Do you want a graphical representation?n
1:Moving disc from A to C
2:Moving disc from A to B
3:Moving disc from A to D
4:Moving disc from B to D
5:Moving disc from A to B
6:Moving disc from D to A
7:Moving disc from D to B
8:Moving disc from A to B
9:Moving disc from C to B
10:Moving disc from A to C
11:Moving disc from A to D
12:Moving disc from C to D
13:Moving disc from A to C
14:Moving disc from D to A
15:Moving disc from D to C
16:Moving disc from A to C
17:Moving disc from A to D
18:Moving disc from C to D
19:Moving disc from C to A
20:Moving disc from D to A
21:Moving disc from C to D
22:Moving disc from A to C
23:Moving disc from A to D
24:Moving disc from C to D
25:Moving disc from B to A
26:Moving disc from B to D
27:Moving disc from B to C
28:Moving disc from D to C
29:Moving disc from B to D
30:Moving disc from C to B
31:Moving disc from C to D
32:Moving disc from B to D
33:Moving disc from A to D
```
Most people will want to answer a single lower case 'y' to the graphics question. The graphics display has the 'A' (source) needle on
the lower left hand side of the display and the Destination needle on the right hand side. There are never any needles above the Destination needle. The whole right hand column except for the destination needle is reserved to move discs vertically to the desired row. This area will be referred to as the 'elevator' In each row there is a little space above the needle for horizontal disc motion. These will be referred to as 'skyways'.

When the graphic dialog asks "How many Columns :" This always excludes the destination needle. On the bottom row there is an extra column. 
The program figures out how many rows it needs. An example needles=8, discs=11, columns=4 will have two rows. The bottom row with 'A', 'B', 'C', 'D' and 'H'.
The top row with 'E', 'F', 'G'.

The next question is "Do you want a faster move of discs?". A 'n' is usually called for. If you want the discs to move in a straight line from where they start to where they are going regardless of what is in the way than a 'y' response is appropriate. This speeds up some puzzles by quite a bit. 

#### The graphics option.

The fast move from A to X is from the top disc on needle A to the place on needle X above the top disc in a straight line.

The normal move is somewhat involved. Where A and X are on same row A's top disc ascends the skyway, moves left or right to just above needle X and 
descends into place.
On different rows A'a top disc ascends to the skyway, moves left to the elevator ascends or descends to the skyway for the row X is on, moves left till it 
is over X and descends into place.

Here is a sample picture captured for the 8 Needle, 11 discs, 3 column example:
![8 Needle, 11 discs, 3 column](N8D11C4.png)

#### Things you might want to try.

In classical Tower of Hanoi with 3 needles and 64 needles, the Sun would become a Red Giant, Earth incinerated and the Milky Way would have collided with Andromeda and you still would not have a solution. But with 4 needles, 64 Discs and no graphics, this takes maybe 7 seconds.


Another to try takes only 127 moves. This is 65 needles, 64 Discs, 8 columns and and fast graphics. This does exactly what you might think. There is no special case for this, it falls out of the more general solution. It is odd the number of moves with 3 needles is 2^n^-1, but with n+1 needles it is 2n-1.

#### For more information.
I highly recommend the following book in either 1st or 2nd edition:  
*The Tower of Hanoi -- Myths and Maths *  
Authored by:  
*Andreas M. Hinz, Sandi Klavzar, Uros Milutinovis, Ciril Petr*  
Published by Birkhauser  
First Edition ISBN 978-3-0348-0237-6  
[Website (unfortunately **unsecure** http) tohbook.info](http://tohbook.info)  
[Widipedia article about the book](https://en.wikipedia.org/wiki/The_Tower_of_Hanoi_%E2%80%93_Myths_and_Maths)


This material is found in a much expanded form in a chapter of a book I have written. I will update this notice when it is published.  
 
[^1]: Reve from the word Reeve: honored official. The Reeve of the Shire is the source of the modern word Sheriff.  
[^2]: Luke 14:28 For which of you, intending to build a tower, sitteth not down first, and counteth the cost, whether he have sufficient to finish it?  


CHANGELOG.txt
Change Log (dates in yyyymmdd format)
==========
0.1.3
- Added 'dan.bates' my PyPi username to the setup.cfg keywords.
  The object is to make it easy for someone to see all my packages.

--------------------------------
0.1.2
- README.md Susing removed, accidental editor command while 0.1.1 changes

0.1.1
- README.md typos corrected

--------------------------------
0.1.0 20220127

- First, Initial Release
 This has been in existence for some months. Just only formalized now and about to go into a git repository and submitted to PyPi.

--------------------------------


src/hanoitoys/HanoiNR.py
#! Python
'''
HanoiNR.py  Tower of Hanoi Non Recursive Solution.

    Classical Hanoi
    Uses move number modulus three to determine Move.
    smallest disc is numbered 0.
'''


def DoMove (i, states, label1, label2):
    # one bit for each disc. bitno is disc number.
    # Determine which direction is legal.
    # states comes in as a tuple and are returned as a tuple.
    state1, state2 = states
    top1 = (state1 & -state1)   # isolate smallest disc top of needle1
    top2 = (state2 & -state2)   # isolate smallest disc top of needle2
    # needle2 empty or (needle1 not empty and smaller disc than needle2).
    if top2 == 0 or (top1 != 0 and top1 < top2):
        state1 &= ~top1  # take disc from top of needle1
        state2 |= top1   # and place it on needle2
        k = top1.bit_length() - 1
        print(f'{i}: Move Disc {k} from {label1} to {label2}')
    else:
        state2 &= ~top2  # take disc from top of needle2
        state1 |= top2   # and place it on needle1
        k = top2.bit_length() - 1
        print(f'{i}: Move Disc {k} from {label2} to {label1}')
    return (state1, state2)  # modified state of two needles


def DoHanoi (n, source, store, destination):
    nMoves = 2 ** n - 1
    if n & 1 == 0:      # when even number of discs relabel needles
        store, destination = destination, store
    #  Source, Temp (not Store), Dest
    #   This algorithm only determines which two needles involved
    #   in next move. Of two possible moves only one is valid.
    S, T, D = nMoves, 0, 0
    action = 0
    for i in range(nMoves):
        action = (action + 1) % 3
        if action == 1:
            S, D = DoMove(i, (S, D), source, destination)
        elif action == 2:
            S, T = DoMove(i, (S, T), source, store)
        elif action == 0:
            T, D = DoMove(i, (T, D), store, destination)
        else:
            assert False


if __name__ == '__main__':
    while True:
        n = int(input("\nHow many discs (0 to halt):"))
        if n == 0:
            break
        DoHanoi(n, 'Source', 'Store', 'Destination')


src/hanoitoys/HanoiClassical.py
#! Python
'''
HanoiClassical.py  Classical Tower of Hanoi Recursive Solution.

    Classical Hanoi
    Base condition n = 1. Just move the disc from source to dest.
    n > 1
    move n-1 from source to store using dest for storage.
    move 1 from source to dest, no store needed.
    move n-1 from store to dest using source for storage.
    smallest disc is numbered 0.
'''


def DoHanoi (n, level, source, store, dest):
    global nmove
    if n == 1:
        print(f'{nmove}:Move Disc {totn - level - 1} from {source} to {dest}')
        nmove += 1
        return
    DoHanoi(n - 1, level + 1, source, dest,   store)
    DoHanoi(1,     level,     source, store,  dest)
    DoHanoi(n - 1, level + 1, store,  source, dest)


def Go ():
    global totn, nmove
    while True:
        n = int(input("\nHow many discs (0 to halt):"))
        if n == 0:
            break
        nmove = 0
        totn = n
        DoHanoi(n, 0, 'Source', 'Store', 'Destination')

if __name__ == '__main__':
    Go()



src/hanoitoys/Hanoi.py
#! Python
'''
Class to represent a Hanoi Tower puzzle. The solving of
Classic Hanoi and Hanoi on steroids (number of needles > 3).
Moves to solve and number of moves calculator est (memoized)

Usage:
    Hanoi(start, set of storage, goal)
examples:
    a = Hanoi('A', OrderedSet(4,1)), 'F', 4)

    experiments have shown 1 << n is much faster than 2 ** n.
    So we use this with the intent pow(2, n)

This has been tested and is equivalent to the Frame-Stewart algorithm.
It was independently developed by self, prior to finding out there
was a Frame-Stewart algorithm.

Note: An earlier version of this used the standard type set
to keep track of available storage needles.
Keeping the available storage needles in a set is problematic
for comparing to Frame-Steward solutions.
A OrderedSet which only holds Natural Numbers defined ISO 80000-2
which includes zero.
'''

import functools    # need @lru_cache for this to work in a reasonable time.
import turtle as T  # Turtle graphics, each disc is a independent turtle.
from tkinter import messagebox

# Greek Letters
# Needles (pegs) labeled A-Z then a-z then alpha-omega

Greek = {'alpha' : '\u03B1', 'beta'    : '\u03B2', 'gamma'   : '\u03B3',
         'delta' : '\u03B4', 'epsilon' : '\u03B5', 'zeta'    : '\u03B6',
         'eta'   : '\u03B7', 'theta'   : '\u03B8', 'iota'    : '\u03B9',
         'kappa' : '\u03BA', 'lambda'  : '\u03BB', 'mu'      : '\u03BC',
         'nu'    : '\u03BD', 'xi'      : '\u03BE', 'omicron' : '\u03BF',
         'pi'    : '\u03C0', 'rho'     : '\u03C1', 'sigma'   : '\u03C2',
         'sigma2': '\u03C3', 'tau'     : '\u03C4', 'upsilon' : '\u03C5',
         'phi'   : '\u03C6', 'chi'     : '\u03C7', 'omega'   : '\u03C7'}

# name space


class Namespace ():
    '''A Namespace is a place to hide variables. The namespace may be global,
    but the contents of the Namespace are not global'''


G = Namespace()  # share data between graphic object and hanoi object


class OrderedSet ():
    '''
    This set only holds non-negative integers. In this case a Needle number.
    The storage for the set is a bit array (like ancient Pascal) stored in
    an int.

    constructor OrderedSet(a, b) sets bits a through b in a new ordered set.
    OrderedSet(a,a) bit with bit number a is set.
    OrderedSet(a,b) with a == b < 0 makes an empty set/
    '''
    def __init__ (self, a, b):
        if a == b < 0:
            self.bitarray = 0
            return
        if a == b:
            self.bitarray = 1 << a
            return
        a, b = sorted([a, b])
        self.bitarray = (1 << b) - (1 << a)

    '''
    os.copy() A shallow copy of the ordered set
    '''
    def copy (self):
        the_copy = OrderedSet(-1, -1)
        the_copy.bitarray = self.bitarray
        return the_copy

    '''
    os.isempty() A predicate returning that there are no elements in os
    '''
    def isempty (self):
        return self.bitarray == 0

    '''
    os.look() returns lowest bit sets number
    and does not modify the bit array
    '''
    def look (self):
        assert not self.isempty()
        k = self.bitarray
        k &= -k
        return k.bit_length() - 1

    '''
    os.pop() returns lowest bit sets number and
    clearing that bit in the bit array.
    '''
    def pop (self):
        assert not self.isempty()
        k = self.bitarray
        k = k & -k
        self.bitarray &= ~k
        return k.bit_length() - 1

    '''
    os.put(n) takes number n and puts it back in the os ordered set.
    '''
    def put (self, n):
        k = 1 << n
        self.bitarray |= k
        return self

    '''
    os.list() returns an ordered list of int s which are the needle
    numbers in os bitarray
    '''
    def list (self):
        x = self.copy()
        L = list()
        while not x.isempty():
            L.append(x.pop())
        return L

    '''
    __len__ another population count how many one bits set in bitarray
    alternatively we could keep track of length. I regard this as safer.
    even though slower.
    '''
    def __len__ (self):
        x = self.bitarray
        population_count = 0
        while x > 0:
            b = x & 0xffffffffffffffff
            x >>= 64
            b -= (b & 0xaaaaaaaaaaaaaaaa) >> 1
            b = ((b & 0xcccccccccccccccc) >> 2) + (b & 0x3333333333333333)
            b = ((b & 0xf0f0f0f0f0f0f0f0) >> 4) + (b & 0x0f0f0f0f0f0f0f0f)
            r = b * 0x0101010101010101
            population_count += (r >> 56) & 0xff
        return population_count


def YorN(question):
    '''
    The Yes or No predicate. Yorn asks a question. Gets a Response
    If the response is affirmate returns True.
    If the reaponse is negative return False.
    Otherwise an error message, list of responses and try again.
    '''
    affirmative = ['affirmative',  'afirmative', 'amen', 'assuredly', 'aye',
                   'certainly', 'definitely', 'exactly', 'fine', 'gladly',
                   'good', 'granted', 'hi', 'indubitably', 'naturally',
                   'okay', 'positively', 'precisely', 'si', 'sure', 'surely',
                   't', 'true', 'undoubtedly', 'unquestionably', 'willingly',
                   'y', 'yah', 'yea', 'yep', 'yes']
    negative =  ['n', 'nay', 'nix', 'never', 'not', 'negative', 'negatory',
                 'nein', 'no', 'noway']
    while True:
        response = input(question).lower().split()
        if response[0] in affirmative:
            return True
        if response[0] in negative:
            return False
        print("This is a yes or no question!")


@functools.lru_cache(maxsize=3000)
def est (NumDiscs: int, NumStore: int) -> tuple:
    '''
    This makes Hanoi on steroids possible. A recursive estimator,
    returns least number of moves and size of pile to get out of
    the way to make this possible.

    Note lru_cache makes this possible. Without that the computation time
    is excessive.

    Number of needles is NumStore + 2.

    idealy this should be a staticmethod of the Hanoi class. However lru_cache
    and static methods mixed together seem to have problems.
    a class method is inappropriate.
    a instance method might have self as part of the cache key, need to
    research this.

    An alternative implementation is not to use a list. Default running best
    to Classic Hanoi to start. And iterate as before. At each iteration, if
    it returns less than running best, then replace running best.
    Finally return running best.
    '''
    # returned tuple (num of moves, first move this many to storage needle
    # can use min on list of these tuples to get tuple to return.
    if NumStore == 1 or NumDiscs < 3:    # base case is classic Hanoi.
        return ((1 << NumDiscs) - 1, 1)  # 2 ** NumDiscs - 1; this is faster.
    possibilities = list()               # build a list of sub problems
    for n in range(1, NumDiscs):
        moves  = est(n, NumStore)[0] << 1           # n src->stor, n stor->dest
        moves += est(NumDiscs - n, NumStore - 1)[0]  # NumDiscs-n src->dest
        possibilities.append((moves, n))  # ok in list see if it is best later
    return min(possibilities)


def info ():
    ''' Show the lru cache information for est()
    Only of interest until cache size tuned.
    '''
    print(f'est.cache_info()={est.cache_info()}')


class Hanoi():

    def __init__ (self        : object,
                  source      : str,  # Discs on this  initially: is name
                  storage     : OrderedSet,  # use these as storage in solution
                  goal        : str,  # Final state all Discs on this needle
                  numberDiscs : int)  ->  object:  # how many discs
        G.storage         = storage.copy()
        G.source          = source
        G.dest            = goal
        G.doGraphics      = False
        G.actualNoofmoves = 0
        G.fast_move       = False
        # ListNeedles maps a Needle number 0 through numberNeedles-1
        #   the appropriate needle name
        G.ListNeedles = [source] + storage.list() + [goal]
        # Top most (smallest) Disc number 0
        # Bottom most (biggest) number numberDiscs - 1
        G.M = dict()    # This is a mapping from needle name to needle number.
        for n, k in enumerate(G.ListNeedles):
            G.M[k] = n
        G.M[source] = 0
        G.towers = list()    # A mapping of needle numbers to Needle objects

    @staticmethod
    def MoveOneDisc (src, dest):
        ''' Perform the operations needed to move the top disc on
        needle named src to needle named dest. including graphics.
        and counting the number of moves.
        This is handles the Base case of a solve_helper()'''
        msrc     =  G.M[src]        # number of needle named by src
        mdest    =  G.M[dest]       # number of needle named by dest
        G.actualNoofmoves += 1      # we have and est. see if it is right.
        print(f'{G.actualNoofmoves}:Moving disc from {LP[src]} to {LP[dest]}')
        if not G.doGraphics:        # if no graphics we are done
            return
        #
        # From here on only needed for Graphics.
        #
        source = G.towers[msrc]     # object needle named by src
        target = G.towers[mdest]    # object needle names by dest
        x0, y0 = source.location    # base location of source needle
        disk = source.remove()      # grab hold of top disc on needle
        target.place(disk)          # place on top of of target needle
        # Now the geometry of moving.
        x1, y1 = target.location    # base location of the dest needle
        ny1 = target.top()[1]       # y component of top of target needle
        # Take disk up to the skyway channel at top free of obstructions.
        if not G.fast_move:
            disk.setposition(x0, y0 + G.skyway)  # up to the skyway
            if source.row != target.row:
                # More moves is a dest needle which is on a different row
                #      than source needle
                # Move disk to the right (the elevator channel)
                disk.setposition(G.elevator, y0 + G.skyway)
                # Move disk up/down to the skyway channel of the elevator
                disk.setposition(G.elevator, y1 + G.skyway)
            # Move disk to the left/right to x position of dest needle
            else:
                # move keeping height
                disk.setposition(x1, y0 + G.skyway)
            if x1 != G.elevator:
                disk.setposition(x1, y1 + G.skyway)
        # Move disk down to top of dest needle
        disk.setposition(x1, ny1)   # if G.fast_move this is it.

    @staticmethod
    def solve_helper (src, stor, dest, nstor, ndiscs):
        ''' the guts of the solution. move n discs to a temporary place,
        move ndiscs-n to  dest. Then move n discs from temporary place
        to dest. n determined by est(...)'''
        if ndiscs == 1:
            Hanoi.MoveOneDisc(src, dest)     # Base Case to stop recursion
            return
        # and another base case. reduce problem to classical.
        if nstor == 1:         # degenerate to regular Hanoi Have only 1 stor.
            where = stor.look()  # no loop just 1 element to worry about
            Hanoi.solve_helper(src,   OrderedSet(dest, dest),
                               where, 1, ndiscs - 1)
            Hanoi.MoveOneDisc(src, dest)
            Hanoi.solve_helper(where, OrderedSet(src, src),
                               dest,  1, ndiscs - 1)
            return

        # Not classic Hanoi, the Frame-Stewart algorithm
        st    = stor.copy()  # do not hurt stor it is a by reference argument
        first = est(ndiscs, nstor)[1]   # always guided by estimator
        last  = ndiscs - first          # size of other stack to move
        first_dest = st.pop()           # get one needle from stor set
        # from src using depleted stor w/dest as storage and final dest as goal
        Hanoi.solve_helper(src, st.copy().put(dest), first_dest, nstor, first)
        # from src using depleted stor as storage agoal is dest
        Hanoi.solve_helper(src, st.copy(),  dest,  nstor - 1, last)
        # from first_dest using depleted stor w/src as storage, dest as goal
        Hanoi.solve_helper(first_dest, st.copy().put(src), dest, nstor, first)

    def solve (self):
        '''
        The member function that does setup for the solution.
        and calls a recursive helper.
        '''
        if G.doGraphics:
            GraphicsSetup()
        G.actualNoofmoves = 0       # a good place to start a counter from.
        # here we separate what has to be recursive.
        self.solve_helper(G.source, G.storage, G.dest,
                          len(G.storage), G.NumberDiscs)
        k = est(G.NumberDiscs, len(G.storage))[0]
        assert G.actualNoofmoves == k, f'predict={k} took={G.actualNoofmoves}'


def redo ():
    '''
    The only thing main calls is redo(). This lets the user execute rerun()
    within an IDE for other test runs.
    '''
    global a, G, L, LP
    while(True):
        G.NumberNeedles = int(input('Enter Number of Needles : '))
        if G.NumberNeedles > 2:
            break
        print('Must have 3 or more Needles!')
    while(True):
        G.NumberDiscs = int(input('Enter Number of Discs : '))
        if G.NumberDiscs > 0:
            break
        print('Number of Discs must be greater than zero!')
    L  = list('ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz')
    LP = L[:]
    for GreekKey, GreekLetter in Greek.items():
        L.append(GreekLetter)
        LP.append(GreekKey)
    src = 0
    dst = G.NumberNeedles - 1
    storage = OrderedSet(1, G.NumberNeedles - 1)
    a = Hanoi(src, storage, dst, G.NumberDiscs)
    G.doGraphics = YorN('Do you want a graphical representation?')
    if G.doGraphics:
        if G.NumberNeedles == 3:
            G.NumberCols = 2
        else:
            while(True):
                G.NumberCols = int(input("How many Columns : "))
                if G.NumberCols > 1:
                    break
                print("Number of Cols must be greater than one")
        G.fast_move = YorN("Do you want a faster move of discs?")

        T.setup(width=1.0, height=1.0, startx=None, starty=None)
        GraphicsPreSetUp()
    a.solve()
    if G.doGraphics:
        messagebox.showinfo("Done", "Press Ok")
        T.bye()

#
# The following is a Turtle Graphics interface
# Turtle graphics will not be explained at this
# time. Turtle Graphics is a Volume 2 topic.
#


class Needle (list):
    "object of class Needle are Hanoi objects to place Discs On"""
    def __init__ (self, number, label, location, row):
        # empty Needle
        self.number   = number      # What is this needles number.
        self.label    = label       # What is this needles name.
        self.location = location    # Where on the Canvas is this needle at.
        self.row      = row         # Which row is needle on.

    def top (self) -> list:
        x, y = self.location
        return [x, y + (len(self) - 1) * G.discthick]

    def place (self, disk):
        self.append(disk)   # push to end of list self

    def remove (self):
        disk = self.pop()   # pop from end of list self
        return disk


class Disc (T.Turtle):
    ''' makings of a disc '''
    def __init__ (self, n):
        # Actual shape is usually a rectangle a resize stretches square.
        T.Turtle.__init__(self, shape='square', visible=False)
        self.penup()
        self.resizemode('user')
        self.shapesize(G.discthick / 20,
                       ((n + 1) / G.NumberDiscs) * G.discwidth / 20)
        colors = ['red', 'orange', 'yellow',
                  'green', 'blue', 'indigo', 'violet']
        el = len(colors)
        self.fillcolor(colors[n % el])
        self.showturtle()


def GraphicsPreSetUp ():
    # don't you just hate it when the setup has a setup
    G.textsize = 20
    G.Nrows = (G.NumberNeedles + G.NumberCols - 2) // G.NumberCols
    t = T.Turtle()
    screen      = t.getscreen()
    G.screenx   = screen.window_width() - 40
    G.screeny   = screen.window_height() - 80
#    G.screenx   = screen.window_width() - 20
#    G.screeny   = screen.window_height() - 40
    xsize = G.screenx // (G.NumberCols + 1)
    ysize = (G.screeny // G.Nrows) // G.NumberDiscs
    G.discwidth = xsize
    G.discthick = min(20, ysize)    # no less than 20, otherwise ysize
    G.skyway = G.discthick * G.NumberDiscs  # skyway: height where discs fly
    del t


def GraphicsSetup ():

    # Needles only need to be drawn once. before we start the solution.
    def DrawNeedle (t, needle, x, y):
        ytop = y + G.NumberDiscs * G.discthick
        t.setposition(x, ytop)
        t.pendown()
        t.setposition(x, y)
        t.penup()
        t.setposition(x - 20, y)
        t.pendown()
        t.setposition(x + 20, y)
        t.penup()
        t.write(needle.label, move=False, align='center',
                font=("Arial", 8, "normal"))

    T.clearscreen()
    t = T.Turtle()
    T.title("Hanoi on steroids.")
    G.height = G.screeny // G.Nrows
    startx = -(G.screenx - G.discwidth) // 2
    starty = -(G.screeny) // 2 + 33
    # elevator is where discs can go up or down
    G.elevator = startx + G.discwidth * G.NumberCols
    for i, needleno in enumerate(G.ListNeedles):
        x = startx + G.discwidth  * (i % G.NumberCols)
        y = starty + G.height * (i // G.NumberCols)
        needle = Needle(i, L[needleno], [x, y], i // G.NumberCols)
        G.towers.append(needle)
        t.hideturtle()
        t.penup()
        if i == len(G.ListNeedles) - 1:
            x, y = G.elevator, starty
            needle.row = 0
        else:
            x, y = needle.location
        DrawNeedle(t, needle, x, y)
        needle.location = [x, y]
    x0, y0 = G.towers[0].location
    for i in range(G.NumberDiscs, 0, -1):
        j = i - 1
        disc = Disc(j)
        disc.penup()
        disc.hideturtle()
        disc.setposition(x0, y0)
        disc.showturtle()
        G.towers[0].place(disc)
        y0 += G.discthick
    messagebox.showinfo("To Start", "Press Ok")
    del t   # t is turtle used to draw needles.

# main


redo()
